File: app/src/main/java/org/dhis2/usescases/datasets/datasetInitial/DataSetInitialActivity.java
package org.dhis2.usescases.datasets.datasetInitial;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;

import com.google.android.material.textfield.TextInputEditText;

import org.dhis2.App;
import org.dhis2.R;
import org.dhis2.commons.Constants;
import org.dhis2.commons.orgunitselector.OUTreeFragment;
import org.dhis2.commons.orgunitselector.OrgUnitSelectorScope;
import org.dhis2.data.dhislogic.CategoryOptionExtensionsKt;
import org.dhis2.data.dhislogic.DhisPeriodUtils;
import org.dhis2.data.dhislogic.OrganisationUnitExtensionsKt;
import org.dhis2.databinding.ActivityDatasetInitialBinding;
import org.dhis2.databinding.ItemCategoryComboBinding;
import org.dhis2.usescases.datasets.dataSetTable.DataSetTableActivity;
import org.dhis2.usescases.general.ActivityGlobalAbstract;
import org.dhis2.utils.category.CategoryDialog;
import org.dhis2.utils.customviews.CategoryOptionPopUp;
import org.dhis2.utils.customviews.PeriodDialog;
import org.dhis2.utils.customviews.PeriodDialogInputPeriod;
import org.hisp.dhis.android.core.category.Category;
import org.hisp.dhis.android.core.category.CategoryOption;
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit;
import org.hisp.dhis.android.core.period.PeriodType;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import javax.inject.Inject;

import kotlin.Unit;

public class DataSetInitialActivity extends ActivityGlobalAbstract implements DataSetInitialContract.View {

    private ActivityDatasetInitialBinding binding;
    View selectedView;
    @Inject
    DataSetInitialContract.Presenter presenter;
    @Inject
    DhisPeriodUtils periodUtils;

    private HashMap<String, CategoryOption> selectedCatOptions;
    private OrganisationUnit selectedOrgUnit;
    private Date selectedPeriod;
    private String dataSetUid;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        dataSetUid = getIntent().getStringExtra(Constants.DATA_SET_UID);
        ((App) getApplicationContext()).userComponent().plus(new DataSetInitialModule(this, dataSetUid)).inject(this);
        super.onCreate(savedInstanceState);

        binding = DataBindingUtil.setContentView(this, R.layout.activity_dataset_initial);
        binding.setPresenter(presenter);
    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.init();
    }

    @Override
    protected void onPause() {
        presenter.onDettach();
        super.onPause();
    }

    @Override
    public void setAccessDataWrite(Boolean canWrite) {

    }

    @Override
    public void setData(DataSetInitialModel dataSetInitialModel) {
        binding.setDataSetModel(dataSetInitialModel);
        binding.catComboContainer.removeAllViews();
        selectedCatOptions = new HashMap<>();
        if (!dataSetInitialModel.categoryComboName().equals("default"))
            for (Category categories : dataSetInitialModel.getCategories()) {
                selectedCatOptions.put(categories.uid(), null);
                ItemCategoryComboBinding categoryComboBinding = ItemCategoryComboBinding.inflate(getLayoutInflater(), binding.catComboContainer, false);
                categoryComboBinding.inputLayout.setHint(categories.displayName());
                categoryComboBinding.inputEditText.setOnClickListener(view -> {
                    selectedView = view;
                    presenter.onCatOptionClick(categories.uid());
                });
                binding.catComboContainer.addView(categoryComboBinding.getRoot());
            }
        else
            presenter.onCatOptionClick(dataSetInitialModel.getCategories().get(0).uid());
        checkActionVisivbility();
    }

    @Override
    public void showOrgUnitDialog(List<OrganisationUnit> data) {
        List<String> preselectedOrgUnits = new ArrayList<>();
        if (selectedOrgUnit != null) {
            preselectedOrgUnits.add(selectedOrgUnit.uid());
        }
        new OUTreeFragment.Builder()
                .showAsDialog()
                .singleSelection()
                .withPreselectedOrgUnits(preselectedOrgUnits)
                .orgUnitScope(new OrgUnitSelectorScope.DataSetCaptureScope(dataSetUid))
                .onSelection(selectedOrgUnits -> {
                    if (!selectedOrgUnits.isEmpty()) {
                        selectedOrgUnit = selectedOrgUnits.get(0);
                        binding.dataSetOrgUnitEditText.setText(selectedOrgUnit.displayName());
                        checkPeriodIsValidForOrgUnit(selectedOrgUnit);
                    }
                    checkActionVisivbility();
                    return Unit.INSTANCE;
                })
                .build()
                .show(getSupportFragmentManager(), OUTreeFragment.class.getSimpleName());
    }

    private void checkPeriodIsValidForOrgUnit(OrganisationUnit selectedOrgUnit) {
        if (selectedPeriod != null && !OrganisationUnitExtensionsKt.inDateRange(selectedOrgUnit, selectedPeriod)) {
            binding.dataSetPeriodEditText.setText(null);
            selectedPeriod = null;
        }
    }

    @Override
    public void showPeriodSelector(PeriodType periodType, List<DateRangeInputPeriodModel> periods, Integer openFuturePeriods) {
        PeriodDialogInputPeriod periodDialog = new PeriodDialogInputPeriod();
        periodDialog.setInputPeriod(periods)
                .setOpenFuturePeriods(openFuturePeriods)
                .setOrgUnit(selectedOrgUnit)
                .setPeriod(periodType)
                .setTitle(binding.dataSetPeriodInputLayout.getHint().toString())
                .setPossitiveListener(selectedDate -> {
                    Calendar calendar = Calendar.getInstance();
                    calendar.setTime(selectedDate);
                    this.selectedPeriod = calendar.getTime();
                    binding.dataSetPeriodEditText.setText(periodUtils.getPeriodUIString(periodType, selectedDate, Locale.getDefault()));
                    checkCatOptionsAreValidForOrgUnit(selectedPeriod);
                    checkActionVisivbility();
                    periodDialog.dismiss();
                })
                .setNegativeListener(v -> {
                    this.selectedPeriod = null;
                    binding.dataSetPeriodEditText.setText(null);
                    checkActionVisivbility();
                })
                .show(getSupportFragmentManager(), PeriodDialog.class.getSimpleName());
    }

    private void checkCatOptionsAreValidForOrgUnit(Date selectedPeriod) {
        int index = 0;
        for (Iterator<Map.Entry<String, CategoryOption>> it = selectedCatOptions.entrySet().iterator(); it.hasNext(); index++) {
            CategoryOption categoryOption = it.next().getValue();
            if (categoryOption != null && !CategoryOptionExtensionsKt.inDateRange(categoryOption, selectedPeriod)) {
                it.remove();
                ((TextInputEditText) binding.catComboContainer.getChildAt(index).findViewById(R.id.input_editText)).setText(null);
            }
        }
    }

    @Override
    public void showCatComboSelector(String catOptionUid, List<CategoryOption> data) {
        if (data.size() == 1 && data.get(0).name().equals("default")) {
            if (selectedCatOptions == null)
                selectedCatOptions = new HashMap<>();
            selectedCatOptions.put(catOptionUid, data.get(0));
        } else if (data.size() <= CategoryDialog.DEFAULT_COUNT_LIMIT) {

            CategoryOptionPopUp.getInstance()
                    .setCategoryName(((TextInputEditText) selectedView).getHint().toString())
                    .setCatOptions(data)
                    .setDate(selectedPeriod)
                    .setOnClick(item -> {
                        if (item != null)
                            selectedCatOptions.put(catOptionUid, item);
                        else
                            selectedCatOptions.remove(catOptionUid);
                        ((TextInputEditText) selectedView).setText(item != null ? item.displayName() : null);
                        checkActionVisivbility();
                    })
                    .show(this, selectedView);
        } else {
            new CategoryDialog(
                    CategoryDialog.Type.CATEGORY_OPTIONS,
                    catOptionUid,
                    true,
                    selectedPeriod,
                    selectedOption -> {
                        CategoryOption categoryOption = presenter.getCatOption(selectedOption);
                        selectedCatOptions.put(catOptionUid, categoryOption);
                        ((TextInputEditText) selectedView).setText(categoryOption != null ? categoryOption.displayName() : null);
                        checkActionVisivbility();
                        return Unit.INSTANCE;
                    }
            ).show(getSupportFragmentManager(),
                    CategoryDialog.Companion.getTAG());
        }
    }

    @Override
    public String getDataSetUid() {
        return dataSetUid;
    }

    @Override
    public OrganisationUnit getSelectedOrgUnit() {
        return selectedOrgUnit;
    }

    @Override
    public Date getSelectedPeriod() {
        return selectedPeriod;
    }

    @Override
    public List<String> getSelectedCatOptions() {
        List<String> selectedCatOption = new ArrayList<>();
        for (int i = 0; i < selectedCatOptions.keySet().size(); i++) {
            CategoryOption catOpt = selectedCatOptions.get(selectedCatOptions.keySet().toArray()[i]);
            selectedCatOption.add(catOpt.uid());
        }
        return selectedCatOption;
    }

    @Override
    public String getPeriodType() {
        return binding.getDataSetModel().periodType().name();
    }

    @Override
    public void setOrgUnit(OrganisationUnit organisationUnit) {
        selectedOrgUnit = organisationUnit;
        binding.dataSetOrgUnitEditText.setText(selectedOrgUnit.displayName());
        binding.dataSetOrgUnitEditText.setEnabled(false);
    }

    @Override
    public void navigateToDataSetTable(String catOptionCombo, String periodId) {
        Bundle bundle = DataSetTableActivity.getBundle(
                dataSetUid,
                selectedOrgUnit.uid(),
                periodId,
                catOptionCombo
        );

        startActivity(DataSetTableActivity.class, bundle, true, false, null);
    }

    private void checkActionVisivbility() {
        boolean visible = true;
        if (selectedOrgUnit == null)
            visible = false;
        if (selectedPeriod == null)
            visible = false;
        for (String key : selectedCatOptions.keySet()) {
            if (selectedCatOptions.get(key) == null)
                visible = false;
        }

        binding.actionButton.setVisibility(visible ? View.VISIBLE : View.GONE);

    }
}

File: app/src/main/java/org/dhis2/usescases/datasets/datasetInitial/DataSetInitialActivity.kt
package org.dhis2.usescases.datasets.datasetInitial

import android.os.Bundle
import android.view.View
import androidx.databinding.DataBindingUtil
import com.google.android.material.textfield.TextInputEditText
import java.util.Calendar
import java.util.Date
import java.util.HashMap
import java.util.Locale
import javax.inject.Inject
import org.dhis2.App
import org.dhis2.R
import org.dhis2.databinding.ActivityDatasetInitialBinding
import org.dhis2.databinding.ItemCategoryComboBinding
import org.dhis2.usescases.datasets.dataSetTable.DataSetTableActivity
import org.dhis2.usescases.general.ActivityGlobalAbstract
import org.dhis2.utils.Constants
import org.dhis2.utils.DateUtils
import org.dhis2.utils.customviews.CategoryOptionPopUp
import org.dhis2.utils.customviews.OrgUnitDialog
import org.dhis2.utils.customviews.PeriodDialog
import org.dhis2.utils.customviews.PeriodDialogInputPeriod
import org.hisp.dhis.android.core.category.CategoryOption
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit
import org.hisp.dhis.android.core.period.PeriodType

class DataSetInitialActivity : ActivityGlobalAbstract(), DataSetInitialView {

    private lateinit var binding: ActivityDatasetInitialBinding

    @Inject
    lateinit var presenter: DataSetInitialPresenter

    private lateinit var selectedView: View

    private var selectedCatOptionsMap: HashMap<String, CategoryOption?>? = null

    private var selectedOrgUnit: OrganisationUnit? = null

    override var selectedPeriod: Date? = null

    private lateinit var dataSetUid: String

    public override fun onCreate(savedInstanceState: Bundle?) {
        dataSetUid = intent.getStringExtra(Constants.DATA_SET_UID)
        (applicationContext as App).userComponent()?.plus(DataSetInitialModule(this, dataSetUid))
            ?.inject(this)
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this, R.layout.activity_dataset_initial)
        binding.presenter = presenter
    }

    override fun onResume() {
        super.onResume()
        presenter.init()
    }

    override fun onPause() {
        presenter.onDettach()
        super.onPause()
    }

    override fun setData(dataSetInitialModel: DataSetInitialModel) {
        binding.dataSetModel = dataSetInitialModel
        binding.catComboContainer.removeAllViews()
        selectedCatOptionsMap = HashMap()
        if (dataSetInitialModel.categoryComboName() != "default") {
            dataSetInitialModel.categories.forEach { category ->
                selectedCatOptionsMap?.put(category.uid(), null)
                val categoryComboBinding = ItemCategoryComboBinding.inflate(
                    layoutInflater,
                    binding.catComboContainer,
                    false
                )
                categoryComboBinding.inputLayout.hint = category.displayName()
                categoryComboBinding.inputEditText.setOnClickListener { view ->
                    selectedView = view
                    presenter.onCatOptionClick(category.uid())
                }
                binding.catComboContainer.addView(categoryComboBinding.root)
            }
        } else {
            presenter.onCatOptionClick(dataSetInitialModel.categories[0].uid())
        }
        checkActionVisivbility()
    }

    private fun clearCatOptionCombo() {
        if (binding.dataSetModel!!.categoryComboName() != "default") {
            repeat(binding.catComboContainer!!.childCount) {
                val catView = binding.catComboContainer!!.getChildAt(it)
                (catView.findViewById<View>(R.id.input_editText) as TextInputEditText).setText(null)
            }
            for (categories in binding.dataSetModel!!.categories) {
                selectedCatOptionsMap!![categories.uid()] = null
            }
        }
    }

    /**
     * When changing orgUnit, date must be cleared
     */
    override fun showOrgUnitDialog(orgUnits: List<OrganisationUnit>) {
        val orgUnitDialog = OrgUnitDialog.getInstace()
        orgUnitDialog
            .setMultiSelection(false)
            .setOrgUnits(orgUnits)
            .setProgram(dataSetUid)
            .setTitle(getString(R.string.org_unit))
            .setPossitiveListener {
                if (!orgUnitDialog.selectedOrgUnit.isNullOrEmpty()) {
                    selectedOrgUnit = orgUnitDialog.selectedOrgUnitModel
                    selectedOrgUnit ?: orgUnitDialog.dismiss()
                    binding.dataSetOrgUnitEditText.setText(selectedOrgUnit!!.displayName())
                    binding.dataSetPeriodEditText.setText(null)
                    selectedPeriod = null
                    clearCatOptionCombo()
                }
                checkActionVisivbility()
                orgUnitDialog.dismiss()
            }
            .setNegativeListener { orgUnitDialog.dismiss() }
            .show(supportFragmentManager, OrgUnitDialog::class.java.simpleName)
    }

    override fun showPeriodSelector(
        periodType: PeriodType,
        periods: List<DateRangeInputPeriodModel>,
        openFuturePeriods: Int?
    ) {
        val periodDialog = PeriodDialogInputPeriod()
        periodDialog.setInputPeriod(periods)
            .setOpenFuturePeriods(openFuturePeriods)
            .setPeriod(periodType)
            .setTitle(binding.dataSetPeriodInputLayout.hint!!.toString())
            .setPossitiveListener { selectedDate ->
                val calendar = Calendar.getInstance()
                calendar.time = selectedDate
                this.selectedPeriod = calendar.time
                binding.dataSetPeriodEditText.setText(
                    DateUtils.getInstance().getPeriodUIString(
                        periodType,
                        selectedDate,
                        Locale.getDefault()
                    )
                )
                clearCatOptionCombo()
                checkActionVisivbility()
                periodDialog.dismiss()
            }
            .setNegativeListener { v ->
                this.selectedPeriod = null
                binding.dataSetPeriodEditText.setText(null)
                checkActionVisivbility()
            }
            .show(supportFragmentManager, PeriodDialog::class.java.simpleName)
    }

    override fun showCatComboSelector(catOptionUid: String, catOptions: List<CategoryOption>) {
        if (catOptions.size == 1 && catOptions[0].name() == "default") {
            selectedCatOptionsMap = selectedCatOptionsMap ?: HashMap()
            selectedCatOptionsMap!![catOptionUid] = catOptions[0]
        } else {
            CategoryOptionPopUp.getInstance()
                .setCategoryName((selectedView as TextInputEditText).hint!!.toString())
                .setCatOptions(catOptions)
                .setDate(selectedPeriod)
                .setOnClick { item ->
                    item?.let {
                        selectedCatOptionsMap!![catOptionUid] = it
                    } ?: selectedCatOptionsMap!!.remove(catOptionUid)

                    (selectedView as TextInputEditText).setText(item?.displayName())
                    checkActionVisivbility()
                }
                .show(this, selectedView)
        }
    }

    override fun getSelectedCatOptions(): List<String> {
        return selectedCatOptionsMap?.values?.map { it?.uid() ?: "" } ?: arrayListOf()
    }

    override fun setOrgUnit(organisationUnit: OrganisationUnit) {
        selectedOrgUnit = organisationUnit
        binding.dataSetOrgUnitEditText.setText(selectedOrgUnit!!.displayName())
        binding.dataSetOrgUnitEditText.isEnabled = false
    }

    override fun navigateToDataSetTable(catOptionCombo: String, periodId: String) {
        val bundle = DataSetTableActivity.getBundle(
            dataSetUid,
            selectedOrgUnit!!.uid(),
            selectedOrgUnit!!.name()!!,
            binding.dataSetModel!!.periodType().name,
            DateUtils.getInstance().getPeriodUIString(
                binding.dataSetModel!!.periodType(),
                selectedPeriod,
                Locale.getDefault()
            ),
            periodId,
            catOptionCombo
        )

        startActivity(DataSetTableActivity::class.java, bundle, true, false, null)
    }

    private fun checkActionVisivbility() {
        val visible = when {
            selectedOrgUnit == null -> false
            selectedPeriod == null -> false
            selectedCatOptionsMap?.values?.any { it == null } ?: true -> false
            else -> true
        }

        binding.actionButton.visibility = if (visible) View.VISIBLE else View.GONE
    }
}
